README.md
# Terminal

## Demo

![Terminal demo](https://cdn.statically.io/gh/bus-stop/x-terminal/master/resources/x-terminal-demo.gif)

## Theme variables

The following theme variables are available to change the colors of the standard theme:

```less
@app-background-color: #000000;
@text-color: #ffffff;
@background-color-selected: #4d4d4d;
@text-color-highlight: #ffffff;
@terminal-color-black: #2e3436;
@terminal-color-red: #cc0000;
@terminal-color-green: #4e9a06;
@terminal-color-yellow: #c4a000;
@terminal-color-blue: #3465a4;
@terminal-color-magenta: #75507b;
@terminal-color-cyan: #06989a;
@terminal-color-white: #d3d7cf;
@terminal-color-bright-black: #555753;
@terminal-color-bright-red: #ef2929;
@terminal-color-bright-green: #8ae234;
@terminal-color-bright-yellow: #fce94f;
@terminal-color-bright-blue: #729fcf;
@terminal-color-bright-magenta: #ad7fa8;
@terminal-color-bright-cyan: #34e2e2;
@terminal-color-bright-white: #eeeeec;
```

## Active Terminal

The active terminal is the terminal that will be used when sending commands to
the terminal.

The active terminal will always have an astrix (`*`) in front of the title.
By default when a terminal is hidden it becomes inactive and the last used
visible terminal will become active. If there are no visible terminals none are
active.

The `Allow Hidden Terminal To Stay Active` setting will change the
default behavior and keep a terminal that is hidden active until another
terminal is focused.

## Services

For plugin writers, the `terminal` package supports two services, `terminal` and `platformioIDETerminal`, which
can be used to easily open terminals. These methods are provided using Atom's [services](http://flight-manual.atom.io/behind-atom/sections/interacting-with-other-packages-via-services/)
API.

To use a service, add a consumer method to consume the service, or
rather a JavaScript object that provides methods to open terminals and run commands.

### 'terminal' service v1.0.0

The `terminal` service provides an [object](https://github.com/atom-community/terminal/blob/29b0751250cb9262fb609db8cae87d87fb383c64/src/terminal.js#L291) with `updateProcessEnv`, `run`, `getTerminalViews`, and `open` methods.

As an example on how to use the provided `run()` method, your
`package.json` should have the following.

```json
{
  "consumedServices": {
    "terminal": {
      "versions": {
        "^1.1.0": "consumePlatformioIDETerminalService"
      }
    }
  }
}
```

Your package's main module should then define a `consumePlatformioIDETerminalService`
method, for example.

```js
import { Disposable } from "atom"

export default {
  terminalService: null,

  consumePlatformioIDETerminalService(terminalService) {
    this.terminalService = terminalService
    return new Disposable(() => {
      this.terminalService = null
    })
  },

  // . . .
}
```

Once the service is consumed, use the `run()` method that is provided
by the service, for example.

```js
// Launch `somecommand --foo --bar --baz` in a terminal.
this.terminalService.run(["somecommand --foo --bar --baz"])
```

# Development

Want to help develop terminal? Here's how to quickly get setup.

First use the [apm](https://github.com/atom/apm) command to clone the
[terminal repo](https://github.com/atom-community/terminal).

```sh
apm develop terminal
```

This should clone the terminal package into the `$HOME/github/terminal`
directory. Go into this directory and install its dependencies.

```sh
cd $HOME/github/terminal
npm install
```

You shouldn't need to rebuild any [node-pty](https://github.com/Tyriar/node-pty)
since they are pre-compiled, however in the event they aren't available,
you can rebuild them with:

```sh
apm rebuild
```

Finally, open this directory in Atom's dev mode and hack away.

```sh
atom --dev
```

There's a test suite available for automated testing of the terminal package.
Simply go to `View > Developer > Run Package Specs` in Atom's main menu or
use the hotkey. You can run the full test suite (which includes running lint
tools) via command-line by running `npm run test` inside the terminal
directory.

Various lint tools are being used to keep the code "beautified". To run only
the lint tools, simply run `npm run lint`.

## Pull Requests

Whenever you're ready to submit a pull request, be sure to submit it
against a fork of the main [terminal repo](https://github.com/atom-community/terminal)
master branch that you'll own. Fork the repo using Github and make note of the
new `git` URL. Set this new git URL as the URL for the `origin` remote in your
already cloned git repo is follows.

```sh
git remote set-url origin ${NEW_GIT_URL}
```

Ensure your new changes passes the test suite by running `npm run test`.
Afterwards, push your changes to your repo and then use Github to submit a new
pull request.

## [xterm.js](https://github.com/xtermjs/xterm.js)

The terminals that users interact with in this package is made possible with
major help from the [xterm.js](https://github.com/xtermjs/xterm.js) library. As
such, often times it's necessary to make changes to xterm.js in order to fix
some bug or implement new features.

If you want to work on xterm.js for the benefit of a bug fix or feature to be
supported in terminal, here's how you can quickly get setup.

First make a fork of [xterm.js](https://github.com/xtermjs/xterm.js). Next,
clone your newly created fork as follows.

```sh
git clone ${YOUR_XTERMJS_FORK} ${HOME}/github/xterm.js
```

Go into your newly cloned repo for xterm.js.

```sh
cd ${HOME}/github/xterm.js
```

Install all needed dependencies.

```sh
npm install
```

Build xterm.js.

```sh
npm run build
```

Ensure the test suite passes.

```sh
npm run test
npm run lint
```

Add a global link for xterm.js to your system.

```sh
npm link
```

Inside your terminal directory, link against the global `xterm` link.

```sh
cd ${HOME}/github/terminal
npm link xterm
```

Finally, perform a rebuild with the [apm](https://github.com/atom/apm) program
inside the terminal directory.

```sh
apm rebuild
```

You're all set for developing xterm.js. Hack away in your xterm.js directory,
run `npm run build`, then reload your Atom window to see the changes to your
terminals.

# Credits and Legal

Click for copyright and license info about this package.

[![LICENSE and © INFO](https://img.shields.io/badge/©%20&%20LICENSE-MIT-blue.svg?longCache=true&=flat-square)](LICENSE)

# Feedback

Need to submit a bug report? Have a new feature you want to see implemented in
_terminal_? Please feel free to submit them through the appropriate
[issue template](https://github.com/atom-community/terminal/issues/new/choose).

For bug reports, please provide images or demos showing your issues if you can.


CHANGELOG.md
## [1.1.8](https://github.com/atom-community/terminal/compare/v1.1.7...v1.1.8) (2022-11-04)


### Bug Fixes

* get correct cwd when treeview is selected ([#102](https://github.com/atom-community/terminal/issues/102)) ([e462b00](https://github.com/atom-community/terminal/commit/e462b00d01cf84060bcb25f976d56ed5d82f1da2))

## [1.1.7](https://github.com/atom-community/terminal/compare/v1.1.6...v1.1.7) (2021-07-25)


### Bug Fixes

* enable scope hoisting of Parcel - reduce the size of app ([00d8d96](https://github.com/atom-community/terminal/commit/00d8d96a7369f783553146b9e4bb092424ba0345))
* require at least Atom 1.52 - electron 6 ([8fc0612](https://github.com/atom-community/terminal/commit/8fc0612973e2f99e74107988853d9c7f97d5734f))
* use terser-config-atomic - reduce the size of the app ([c17c299](https://github.com/atom-community/terminal/commit/c17c29934b80670b8f7a091d2be579c4d52ad770))

## [1.1.6](https://github.com/atom-community/terminal/compare/v1.1.5...v1.1.6) (2021-07-13)


### Bug Fixes

* use padding variable ([81cf0c9](https://github.com/atom-community/terminal/commit/81cf0c96c54bf8cec4ca58f2e8689d867a04c46c))

## [1.1.5](https://github.com/atom-community/terminal/compare/v1.1.4...v1.1.5) (2021-07-12)


### Bug Fixes

* add padding to the terminal ([14144a5](https://github.com/atom-community/terminal/commit/14144a53e6055c18b9d0855a3d14b4d20101c4d7))
* add terminal syntax variables to standard theme ([bb27e56](https://github.com/atom-community/terminal/commit/bb27e56689e271b9216ef0039ab1aeccbcc52eaf))
* default to Standard theme ([512e285](https://github.com/atom-community/terminal/commit/512e28597aa4bd5d72214604e8b053ace3cb067f))
* update Dependencies ([a9e6be8](https://github.com/atom-community/terminal/commit/a9e6be82fef424d68686dd56189dbf156996c1f4))
* update theme without restarting terminal ([b38d4d1](https://github.com/atom-community/terminal/commit/b38d4d18b4ae5ce87598c11eeb51ec31dbcc0cd5))
* use 1rem of padding ([834a710](https://github.com/atom-community/terminal/commit/834a710537884fb0d92d40d059ef514bbbd1798d))

## [1.1.4](https://github.com/atom-community/terminal/compare/v1.1.3...v1.1.4) (2021-06-13)


### Bug Fixes

* update Dependencies ([d1f3d5d](https://github.com/atom-community/terminal/commit/d1f3d5d1570337f1a60f5b05245979511db3b627))

## [1.1.3](https://github.com/atom-community/terminal/compare/v1.1.2...v1.1.3) (2021-05-13)


### Bug Fixes

* update dependencies ([dcae454](https://github.com/atom-community/terminal/commit/dcae45439510fcb62d88044fa65450f13f63a842))

## [1.1.2](https://github.com/atom-community/terminal/compare/v1.1.1...v1.1.2) (2021-05-08)


### Bug Fixes

* use customElements ([bcf29f0](https://github.com/atom-community/terminal/commit/bcf29f0118f8a7d81a1a34b582a91063314aed62))

## [1.1.1](https://github.com/atom-community/terminal/compare/v1.1.0...v1.1.1) (2021-05-07)


### Bug Fixes

* toolbar config was wrong ([08625fb](https://github.com/atom-community/terminal/commit/08625fb3c83c586ec384ca2e9f2535c4958c03c4))
* use node url instead of whatwg-url ([71de4dc](https://github.com/atom-community/terminal/commit/71de4dc495f244257de8887fd860f5df9929f2ee))

# [1.1.0](https://github.com/atom-community/terminal/compare/v1.0.1...v1.1.0) (2021-05-03)


### Features

* add setting to remove toolbar button ([6b66057](https://github.com/atom-community/terminal/commit/6b66057c5a0ea4590d5d595a9f03f46fbc484e32))

## [1.0.1](https://github.com/atom-community/terminal/compare/v1.0.0...v1.0.1) (2021-05-03)


### Bug Fixes

* activate pane on focus ([cee1511](https://github.com/atom-community/terminal/commit/cee1511f594467a1601f96c43fd1645bb4216f57))

# 1.0.0 (2021-04-29)


### Bug Fixes

* allow showing the terminal when intersectionRatio is not 1 ([7a05451](https://github.com/atom-community/terminal/commit/7a0545149fc5d5917a322bd06f9f9af7a5308717))
* bump deps ([4de3aea](https://github.com/atom-community/terminal/commit/4de3aea97b0a0495b583c6cf90a63762032bdb7f))
* close on error opening shell ([11225e9](https://github.com/atom-community/terminal/commit/11225e9d4f5d72cb4c28338bed3f88434e5f9d73))
* enable ligatures ([10e64ce](https://github.com/atom-community/terminal/commit/10e64ceba8f13eadc96299da526a59d16d1e341d))
* export getDefaultShell and getFallbackShell for testing ([f8fa49b](https://github.com/atom-community/terminal/commit/f8fa49ba7dec189f93f24055974f642e25a09e19))
* Fix context menu on terminal ([#38](https://github.com/atom-community/terminal/issues/38)) ([f9a528e](https://github.com/atom-community/terminal/commit/f9a528e07b6b2a0d084f9544d4206a05bcbcf864))
* move watcher disposable code ([d8b4fc9](https://github.com/atom-community/terminal/commit/d8b4fc9bb5fab14ed77776ff5d0533ce704fc331)), closes [/github.com/atom-community/terminal/pull/24#discussion_r499099903](https://github.com//github.com/atom-community/terminal/pull/24/issues/discussion_r499099903)
* set autoShell only for the first time ([977e6dc](https://github.com/atom-community/terminal/commit/977e6dc528e557d785587fc5dd9e2d7a0930f482))
* set shell to pwsh or powershell if exists ([60b28f4](https://github.com/atom-community/terminal/commit/60b28f414471aa6a2753ca7b8bc1072c29f10eda))
* typo defaultShellCache ([151d6b2](https://github.com/atom-community/terminal/commit/151d6b2640a291dff0f498c2ef65f3976b31dda8))
* update deps ([3944561](https://github.com/atom-community/terminal/commit/3944561566f501d759b3457d26534a12e8a4474b))
* update deps ([ec03e67](https://github.com/atom-community/terminal/commit/ec03e673b6f001a955b2d6e2f2ea7021dcd6c64c))
* upgrade whatwg-url from 8.3.0 to 8.4.0 ([a488877](https://github.com/atom-community/terminal/commit/a488877a2a2d1faaf86c7034e998feee9cc81d8e))
* use async exists ([02ad971](https://github.com/atom-community/terminal/commit/02ad9714d15ffe09c13b5a4e982946da262cefc8))
* use fallback in case of an error ([c168765](https://github.com/atom-community/terminal/commit/c168765a9a0190332414992c6edc73bb3703b89f))

src/config.ts
type configObjects = Record<string, configObject>

type configObject = {
  title: string
  description: string
  default?: string | boolean | number
  enum?: (string | number)[]
  minimum?: number
  maximum?: number
  order?: number
  collapsed?: boolean
  type: string
  properties?: configObjects
}

function configOrder(obj: configObjects): configObjects {
  let order = 1
  for (const name in obj) {
    obj[name].order = order++
    if (obj[name].type === "object" && "properties" in obj[name]) {
      configOrder(<configObjects>obj[name].properties)
    }
  }
  return obj
}

export function getDefaultShell(): string {
  return process.platform === "win32" ? process.env.COMSPEC || "cmd.exe" : process.env.SHELL || "/bin/sh"
}

export const config = configOrder({
  shell: {
    title: "Shell",
    description: "Path to the shell command.",
    type: "string",
    default: getDefaultShell(),
  },
  encoding: {
    title: "Character Encoding",
    description: "Character encoding to use in spawned terminal.",
    type: "string",
    default: "",
  },
  webgl: {
    title: "WebGL Renderer",
    description: "Enable the WebGL-based renderer.",
    type: "boolean",
    default: false,
  },
  webLinks: {
    title: "Web Links",
    description: "Enable clickable web links.",
    type: "boolean",
    default: true,
  },
  ligatures: {
    title: "Ligatures",
    description: "Enables ligatures support.",
    type: "boolean",
    default: true,
  },
  fontFamily: {
    title: "Font Family",
    description: "Font family used in terminal emulator.",
    type: "string",
    default: atom.config.get("editor.fontFamily") || "monospace",
  },
  fontSize: {
    title: "Font Size",
    description: "Font size used in terminal emulator.",
    type: "integer",
    default: atom.config.get("editor.fontSize") || 14,
    minimum: 8,
    maximum: 100,
  },
  defaultOpenPosition: {
    title: "Default Open Position",
    description: "Position to open terminal through service API or terminal:open.",
    type: "string",
    enum: ["Center", "Split Up", "Split Down", "Split Left", "Split Right", "Bottom Dock", "Left Dock", "Right Dock"],
    default: "Bottom Dock",
  },
  allowHiddenToStayActive: {
    title: "Allow Hidden Terminal To Stay Active",
    description: "When an active terminal is hidden keep it active until another terminal is focused.",
    type: "boolean",
    default: false,
  },
  runInActive: {
    title: "Run in Active Terminal",
    description: "Whether to run commands from the service API in the active terminal or in a new terminal.",
    type: "boolean",
    default: true,
  },
  allowRelaunchingTerminalsOnStartup: {
    title: "Allow relaunching terminals on startup",
    description: "Whether to allow relaunching terminals on startup.",
    type: "boolean",
    default: false,
  },
  copyOnSelect: {
    title: "Copy On Select",
    description: "Copy text to clipboard on selection.",
    type: "boolean",
    default: true,
  },
  toolbarButton: {
    title: "Add Toolbar Button",
    description: "Add button to open a terminal from the toolbar.",
    type: "boolean",
    default: true,
  },
  colors: {
    title: "Colors",
    description: "Settings for the terminal colors.",
    type: "object",
    collapsed: true,
    properties: {
      theme: {
        title: "Theme",
        description: "Theme used in terminal emulator.",
        type: "string",
        enum: [
          "Custom",
          "Atom Dark",
          "Atom Light",
          "Base16 Tomorrow Dark",
          "Base16 Tomorrow Light",
          "Christmas",
          "City Lights",
          "Dracula",
          "Grass",
          "Homebrew",
          "Inverse",
          "Linux",
          "Man Page",
          "Novel",
          "Ocean",
          "One Dark",
          "One Light",
          "Predawn",
          "Pro",
          "Red Sands",
          "Red",
          "Silver Aerogel",
          "Solarized Dark",
          "Solarized Light",
          "Solid Colors",
          "Standard",
        ],
        default: "Standard",
      },
      // TODO: add transparency settings?
      foreground: {
        title: "Text Color",
        description: "This will be overridden if the theme is not 'Custom'.",
        type: "color",
        default: "#ffffff",
      },
      background: {
        title: "Background Color",
        description: "This will be overridden if the theme is not 'Custom'.",
        type: "color",
        default: "#000000",
      },
      cursor: {
        title: "Cursor Color",
        description: "Can be transparent. This will be overridden if the theme is not 'Custom'.",
        type: "color",
        default: "#ffffff",
      },
      cursorAccent: {
        title: "Cursor Text Color",
        description: "Can be transparent. This will be overridden if the theme is not 'Custom'.",
        type: "color",
        default: "#000000",
      },
      selection: {
        title: "Selection Background Color",
        description: "Can be transparent. This will be overridden if the theme is not 'Custom'.",
        type: "color",
        default: "#4d4d4d",
      },
      black: {
        title: "ANSI Black",
        description: "`\\x1b[30m`",
        type: "color",
        default: "#2e3436",
      },
      red: {
        title: "ANSI Red",
        description: "`\\x1b[31m`",
        type: "color",
        default: "#cc0000",
      },
      green: {
        title: "ANSI Green",
        description: "`\\x1b[32m`",
        type: "color",
        default: "#4e9a06",
      },
      yellow: {
        title: "ANSI Yellow",
        description: "`\\x1b[33m`",
        type: "color",
        default: "#c4a000",
      },
      blue: {
        title: "ANSI Blue",
        description: "`\\x1b[34m`",
        type: "color",
        default: "#3465a4",
      },
      magenta: {
        title: "ANSI Magenta",
        description: "`\\x1b[35m`",
        type: "color",
        default: "#75507b",
      },
      cyan: {
        title: "ANSI Cyan",
        description: "`\\x1b[36m`",
        type: "color",
        default: "#06989a",
      },
      white: {
        title: "ANSI White",
        description: "`\\x1b[37m`",
        type: "color",
        default: "#d3d7cf",
      },
      brightBlack: {
        title: "ANSI Bright Black",
        description: "`\\x1b[1;30m`",
        type: "color",
        default: "#555753",
      },
      brightRed: {
        title: "ANSI Bright Red",
        description: "`\\x1b[1;31m`",
        type: "color",
        default: "#ef2929",
      },
      brightGreen: {
        title: "ANSI Bright Green",
        description: "`\\x1b[1;32m`",
        type: "color",
        default: "#8ae234",
      },
      brightYellow: {
        title: "ANSI Bright Yellow",
        description: "`\\x1b[1;33m`",
        type: "color",
        default: "#fce94f",
      },
      brightBlue: {
        title: "ANSI Bright Blue",
        description: "`\\x1b[1;34m`",
        type: "color",
        default: "#729fcf",
      },
      brightMagenta: {
        title: "ANSI Bright Magenta",
        description: "`\\x1b[1;35m`",
        type: "color",
        default: "#ad7fa8",
      },
      brightCyan: {
        title: "ANSI Bright Cyan",
        description: "`\\x1b[1;36m`",
        type: "color",
        default: "#34e2e2",
      },
      brightWhite: {
        title: "ANSI Bright White",
        description: "`\\x1b[1;37m`",
        type: "color",
        default: "#eeeeec",
      },
    },
  },
})

// finds the default shell start commmand
export async function setAutoShell(whichFun: typeof which): Promise<void> {
  let shellStartCommand
  if (process.platform === "win32") {
    // Windows
    try {
      shellStartCommand = await whichFun("pwsh.exe")
    } catch (e1) {
      try {
        shellStartCommand = await whichFun("powershell.exe")
      } catch (e2) {
        // keep default
      }
    }
  }

  if (shellStartCommand && atom.config.get("atomic-terminal.shell") === getDefaultShell()) {
    atom.config.set("atomic-terminal.shell", shellStartCommand)
  }
}

import which from "which"

// set shell command automatically on first install
if (localStorage.getItem("atomic-terminal.autoShellSet") === null) {
  localStorage.setItem("atomic-terminal.autoShellSet", "true")
  setAutoShell(which)
}
